### PR TITLE
Radioactive decay for RTGs

### PR DESCRIPTION
**Pull request by deltanedas**
## About the PR
- damaged rtg's radiation has a half life of 5 minutes
- added RadiationDecayComponent and SharedRadiationSystem
- moved RadiationSource modification into an api provided by the latter
- behonker emits radiation to fix test

if you want to do suppermatter its an emergency thing and not a roundstart permanent thing (marathon)
or you will need to get fresh supply of rtgs from salvage

**Media**
- [X] I have added screenshots/videos to this PR showcasing its changes ingame, **or** this PR does not require an ingame showcase

**Changelog**
:cl:
- tweak: Damaged RTG's radiation now weakens over time.

### CODE REVIEW

File: Content.Shared/Radiation/Systems/SharedRadiationSystem.cs [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+
+    /// <summary>
+    /// Set the Intensity field of a radiation source
+    /// </summary>
+    public void SetIntensity(EntityUid uid, float intensity, RadiationSourceComponent? comp = null)
+    {
+        if (!Resolve(uid, ref comp))
+            return;
+
+        comp.Intensity = intensity;
```
Comment (deltanedas) [inline Content.Shared/Radiation/Systems/SharedRadiationSystem.cs:28]: idk if these should == and check dirty

**Comment by Interrobang01** [comment]
Suppermatter? Don't you mean supermatter?

Anyways this looks cool although a halflife of 5 minutes seems a bit short, but I suppose smashing an rtg and building a slightly larger rtg around it shouldn't work that well.

**Comment by EmoGarbage404** [comment]
i feel like this is an issue with radiation collectors being basically free energy for any source of radiation. There's nothing stopping someone from doing the same thing with a rad artifact or any other source of radiation.

Perhaps the output of radiation collectors just needs to be nerfed so that a single RTG isn't enough to power a whole station. The power output already scales with the rad amount, so just balance it so that singulos remain an effective power source while other sources are too weak.

On just a general game design perspective, the rads of a structure dropping so rapidly seems very unintuitive. Especially for RTGs harvested on salvage and hidey-holes, they might have barely an radiation left by the time someone gets to them, which could feel really weird.

**Comment by deltanedas** [comment]
they dont decay while intact only after breaking

**Comment by Anzarot121** [comment]
I think 5 minutes is too short. Maybe 10 or 15 minutes would be fine.

**Comment by lzk228** [comment]
1984

**Comment by Interrobang01** [comment]
making the rtg have decay makes sense from a gameplay perspective since smashing something shouldn't make it better, and it kind of makes sense from a realism perspective (all the juice is leaking out) unless you think about it too hard

really 5 minutes is a good half-life, smashing your power generators should be more threatening than helpful

**Comment by deltanedas** [comment]
remember 5 mins is halflife not lifetime it just falls off

**Comment by retequizzle** [comment]
issue could also just be addressed by porting the AL generator fuel system to the RTG so that it requires fuel, because as is it doesn't require any

**Comment by Anzarot121** [comment]
> issue could also just be addressed by porting the AL generator fuel system to the RTG so that it requires fuel, because as is it doesn't require any

This is fucking point of RTG. It doesn't have to require fuel, rather, it works by decaying the nuclear fuel inside it, and then converting that decay heat into electricity.

**Comment by deltanedas** [comment]
yeah rtg is meant to run for a very long time unattended
by damaging it you are speeding it up letting you potentially make more power but it is no longer infinite

**Review by Partmedia — COMMENTED** [review]
Consider:
- Making the radiation decay part of the `RadiationSource` component
- Storing a "half life" data field in YAML instead of a rate; a rate is much harder to understand and balance
- Making the half life quite a bit longer than 5 minutes for the reasons stated above

You might look at https://github.com/Partmedia/mining-station-14/commit/f7b88f37140755f52ad88f01700f49d1466645fa for some ideas.

**Comment by deltanedas** [comment]
putting decay on source directly is bad since artifacts, grav anomalies, singulos do not decay
having half life in logic would be bad for performance (no benchmark :trollface:) would Rate get; set {calculate from halflife} be ok for that
and ill change to 15 mins

**Comment by OctoRocket** [comment]
Smashing an RTG should only be a source of radiation, not of power. It should significantly reduce the power output, even with radiation collectors.

**Comment by Partmedia** [comment]
> putting decay on source directly is bad since artifacts, grav anomalies, singulos do not decay having half life in logic would be bad for performance (no benchmark :trollface:) would Rate get

The performance claims are baloney. Components are objects and so the cache miss from accessing a different component, plus *running a whole new entity system*, would swamp the half life check. With this implementation you have to TryComp to make sure that radioactive decay components also have a radiation source component, which is not free.

So the only real question is whether not representing something that does not decay with say, zero half life is acceptable. It is to me but some others might disagree and say that separating them into separate components is worth the extra overhead.

Regardless I insist that "half life" is the value stored in the prototype, not "rate".

**Comment by deltanedas** [comment]
yeah so would a get; set {} thing be correct for that

**Comment by deltanedas** [comment]
i got the math wrong :trollface:

**Comment by Partmedia** [comment]
Setting half life in YAML and combining it into RadiationSource makes much more sense, thank you.

I'm curious why RadiationSource is now being networked and shared? In theory, doesn't this permit a client without a Geiger counter to still be able to identify things that are radioactive?

**Comment by deltanedas** [comment]
i thought the client would mispredict if it wasnt told somethings decay rate, is that wrong

**Comment by deltanedas** [comment]
ok i am abandoning this since the deltaTime math is trolling me
ticking twice in same second should have same effect as ticking once in a second

the current approach will not decay enough
a better approach decays slightly too much at high dt, and way too much at low dt

**Comment by Partmedia** [comment]
> ok i am abandoning this since the deltaTime math is trolling me ticking twice in same second should have same effect as ticking once in a second
> 
> the current approach will not decay enough a better approach decays slightly too much at high dt, and way too much at low dt 
> 
> this is just 10 tps and it is wildly inaccurate 

I think this is salvageable. You can keep the existing difference equation and it'll be "accurate enough" as long as `dt` is small, and even if the error is big, as long as you can tune it to be reasonably balanced it'll be fine.

Alternatively, you save the time of component init as a timespan (wrt gameticking), then every tick, you get the current time, subtract the time difference, and calculate using the theoretical equation what the current fraction is supposed to be. This will give you an exact answer.

**Comment by YotaXP** [comment]
Give this a try.
`comp.Intensity *= MathF.Exp(-comp.DecayRate * frameTime)`

Recommended viewing for this stuff: https://www.youtube.com/watch?v=LSNQuFEDOyQ

**Pull request closed without merging**